File: psBLUP-main/psBLUP.R
### psBLUB ----
psBLUP <- function(
  Y = NULL,
  X = NULL,
  proximityMatrix = NULL,
  logTransformX = FALSE,
  percentTrainData = NULL,
  l2s = seq(from = 1, to = 75, length.out = 7),
  yVariables = NULL,
  xVariables = NULL,
  groupingVariables = NULL,
  runs = 5,
  compareToRRBLUP = T
) {
  ##############################################################################
  # Arguments:
  # Y: data.frame of the response variables
  # X: data.frame of the predictor variables
  # proximityMatrix: A matrix with the proximity between predictor variables.
  #                  If its NULL, the observed correlation between the X
  #                  variables will be used
  # logTransformX: TRUE/FALSE on whether a log-transformation needs to be
  #                implemented on the response variables
  # percentTrainData: A vector with the percentages to be used for
  #                   training the model
  # l2s: A vector of penalties for inducing shrinkage on the differences between
  #      predictor variables
  # yVariables: a vector of names for the response variables to be used
  # xVariables: a vector of names for the predictor variables to be used
  # groupingVariables: a vector of names for variable included in the predictor
  #                    variables that we need to remove their effect, e.g.
  #                    treatment effects
  # runs: number of runs for calculating the final accuracy
  # compareToRRBLUP: TRUE/FALSE on whether the results need to be compared to
  #                  RRBLUP
  ##############################################################################

  ### Load relevant packages | were loaded before
  packages <- c("data.table", "plyr", "rrBLUP")
  lapply(packages, require, character.only = TRUE)

  ### Do the necessary checks for arguments
  if (is.null(X)) {
    stop("No predictor variables have been provided")
  }
  if (is.null(Y)) {
    stop("No dependent variables have been provided")
  }
  if (is.null(X)) {
    proximityMatrix <- cor(X)
  }
  if (!logTransformX %in% c(T, F)) {
    stop("Acceptable values for 'logTransformX': 'TRUE' or 'FALSE'")
  }
  if (!compareToRRBLUP %in% c(T, F)) {
    stop("Acceptable values for 'compareToRRBLUP': 'TRUE' or 'FALSE'")
  }
  if (is.null(yVariables)) {
    yVariables <- names(Y)
  }
  if (is.null(xVariables)) {
    xVariables <- names(X)
  }
  if (is.null(percentTrainData)) {
    percentTrainData <- 1
    runs = 1
    warning(
      "Since no training data will be used, only 1 run will be conducted."
    )
  }
  objectsToReturn <- list()
  ### Transform data to data.table format
  setDT(X, keep.rownames = TRUE)
  setDT(Y, keep.rownames = TRUE)

  ### Keep only relevant variables
  Y <- Y[, c("rn", yVariables, groupingVariables), with = F]
  ### Keep only relevant variables
  X <- X[, c("rn", xVariables), with = F]

  ### Rename percentTrainData to proportion
  proportion = percentTrainData
  ### Apply log transformation to the data
  if (logTransformX) {
    Y[,
      c(yVariables) := lapply(.SD, function(x) {
        log(x)
      }),
      .SDcols = yVariables
    ]
  }

  ### Keep mean level of dependent variables per grouping variable
  ### Also remove grouping effect by regressing them out
  if (!is.null(groupingVariables)) {
    meanByGroup <- copy(Y)
    f <- as.formula(paste(
      "x",
      paste(groupingVariables, collapse = "+"),
      sep = "~"
    ))
    meanByGroup[,
      c(yVariables) := lapply(.SD, function(x) {
        aov(formula = x ~ Treatment)$fitted.values
      }),
      .SDcols = yVariables
    ]
    Y[,
      c(yVariables) := lapply(.SD, function(x) {
        aov(formula = x ~ Treatment)$resid
      }),
      .SDcols = yVariables
    ]
    meanByGroup[,
      c(groupingVariables) := lapply(.SD, function(c) {
        NULL
      }),
      .SDcols = groupingVariables
    ]
    Y[,
      c(groupingVariables) := lapply(.SD, function(c) {
        NULL
      }),
      .SDcols = groupingVariables
    ]
  }

  ### Add strength (row-sum) as diagonal (needed for properly defining the Laplacian matrix)
  diag(proximityMatrix) <- apply(proximityMatrix, 1, sum)
  ### Scale the X variables
  X[,
    c(xVariables) := lapply(.SD, function(x) {
      scale(x)
    }),
    .SDcols = xVariables
  ]

  ### Define N as the number of samples
  N <- nrow(X)

  ### Run psBLUP for every metabolite
  for (yVar in yVariables) {
    ### define as "response" the scaled concentration  of the selected metabolite
    response <- c(scale(Y[, yVar, with = F]))
    ### define as "predictors" the SNP data
    predictors <- copy(X)
    ### Find which prediction variables that are negatively correlated to the response so that
    ### neighboring SNPs with opposing effects will not negate true signal
    negativeCorrelations <- which(
      cor(response, predictors[, xVariables, with = F]) < 0
    )
    neg.effects <- xVariables[negativeCorrelations]
    predictors[,
      c(neg.effects) := lapply(.SD, function(x) {
        a <- min(x)
        b <- max(x)
        return(mapvalues(x, from = c(a, b), to = c(b, a)))
      }),
      .SDcols = neg.effects
    ]

    ### Define the Normalized Laplacian matrix as G
    G <- proximityMatrix
    diagG <- diag(G)
    G <- (-G)
    diag(G) <- diagG
    ### Decompose the Normalize Laplacian matrix and keep the eigen-vectors and eigen-values
    eigenvalues <- eigen(G)$values
    eigenvalues[which(eigenvalues < 0)] <- 0
    eigenvectors <- eigen(G)$vectors
    ### Define as "newdata" the part that will be used for the augmented data solution
    newdata <- t(eigenvectors %*% diag(sqrt(eigenvalues)))
    colnames(newdata) <- rownames(newdata) <- xVariables
    ### Create a matrix to store all differences between RRBLUP and psBLUP per run
    accuracy <- matrix(NA, length(proportion), runs)
    colnames(accuracy) <- paste0("run: ", 1:runs)
    rownames(accuracy) <- paste0("train sample: ", proportion * 100, "%")

    ### Create a matrix containing the final accuracy results
    ### which will be the aggregated results over all runs
    overallAccuracy <- matrix(
      0,
      length(proportion),
      length(l2s) + compareToRRBLUP
    )
    colnames(overallAccuracy) <- c(
      if (compareToRRBLUP) {
        "RRBLUP"
      },
      paste0("l2 = ", signif(l2s, digits = 3))
    )
    rownames(overallAccuracy) <- paste0(
      "train = ",
      signif(proportion, digits = 2)
    )

    ### Run accuracy evaluation using while, as the samples might not contain genetic variation
    ### and produce error in fitting the models
    j <- 0
    while (j < runs) {
      ### Create a matrix for storing the accuracy per sampling scenario
      accuracyPerScenario <- matrix(
        NA,
        length(proportion),
        length(l2s) + compareToRRBLUP
      )
      colnames(accuracyPerScenario) <- c(
        if (compareToRRBLUP) {
          "RRBLUP"
        },
        paste0("l2 = ", signif(l2s, digits = 3))
      )
      rownames(accuracyPerScenario) <- paste0(
        "train = ",
        signif(proportion, digits = 2)
      )
      for (k in 1:length(proportion)) {
        ### Select the training and testing data
        train <- sample(N, round(proportion[k] * N))
        test <- c(1:N)[-train]
        X.train <- predictors[train, ]
        Y.train <- response[train]
        if (length(train) == N) {
          X.test <- X.train
          Y.test <- Y.train
        } else {
          X.test <- predictors[test, ]
          Y.test <- response[test]
        }

        if (compareToRRBLUP) {
          ### Fit RRBLUP
          mixed.new <- mixed.solve(
            Y.train,
            Z = X.train[, xVariables, with = F],
            method = "REML"
          )
          ### Obtain the RRBLUP coefficients
          coeffsRRBLUP <- mixed.new$u
          ### Get fitted values of RRBLUP
          fittedRRBLUP <- c(
            as.matrix(X.test[, xVariables, with = F]) %*% coeffsRRBLUP
          ) +
            mixed.new$beta
          ### Calculate and store RRBLUP accuracy on the correct cell
          accuracyPerScenario[k, 1] <- signif(
            cor(Y.test, fittedRRBLUP),
            digits = 3
          )
        }
        ### Run psBLUP for every l2 value
        for (i in 1:length(l2s)) {
          l2 = l2s[i]
          ### Calculate the factor to be premultiplied to the data for correctly defining the augmented data solution
          factor <- (1 + l2)^(-1 / 2)
          ### Create the augmented data for the predictors and response
          Xdata <- (factor *
            (rbind(X.train[, xVariables, with = F], sqrt(l2) * newdata)))
          Ydata <- (c(Y.train, rep(0, times = nrow(newdata))))
          Xdata <- data.frame(Xdata)
          ### Fit psBLUP
          mixed.new <- tryCatch(
            mixed.new <- mixed.solve(Ydata, Z = Xdata, method = "REML"),
            error = function(e) NA
          )
          ### Check if some model components have not been defined correctly
          if (sum(is.na(mixed.new)) == 1) {
            break
          } else {
            ### Obtain the psBLUP coefficients
            coeffsPSBLUP <- mixed.new$u
            ### Obtain the fitted values
            fittedPSBLUP <- c(
              as.matrix(X.test[, xVariables, with = F]) %*% coeffsPSBLUP
            ) +
              mixed.new$beta
            ### Calculate the psBLUP accuracy for a specific l2 and store it in the appropriate cell
            accuracyPerScenario[k, i + compareToRRBLUP] <- signif(
              cor(Y.test, fittedPSBLUP),
              digits = 3
            )
          }
        }
        ### For a specific scenario of training and testing data,
        ### the difference between RRBLUP and the max accuracy observed for a specific l2 is the accuracy gain
        accuracy[k, j + 1] <- max(accuracyPerScenario[
          k,
          (compareToRRBLUP + 1):ncol(accuracyPerScenario)
        ]) -
          accuracyPerScenario[k, 1] * compareToRRBLUP
      }
      j <- j + 1
      overallAccuracy <- overallAccuracy + accuracyPerScenario
    }
    overallAccuracy <- overallAccuracy / j
    objectsToReturn[[yVar]] <- list(
      "accuracy" = overallAccuracy,
      "gain" = accuracy
    )
  }
  # to extract the accuracy of psBLUP and RRBLUP you can use the following code:
  # accuracy_ps <- objectsToReturn[[1]]$accuracy[1,1]
  # accuracy_rr <- objectsToReturn[[1]]$accuracy[1,2]
  # accuracy_gain <- objectsToReturn[[1]]$gain
  return(objectsToReturn)
}

File: psBLUP.py
import warnings
from typing import Dict, List, Optional

from tqdm import tqdm
import numpy as np
import pandas as pd
from scipy import linalg
from scipy.stats import pearsonr
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import StandardScaler


def psblup(
    Y: Optional[pd.DataFrame] = None,
    X: Optional[pd.DataFrame] = None,
    proximity_matrix: Optional[np.ndarray] = None,
    log_transform_x: bool = False,
    percent_train_data: Optional[float] = None,
    l2s: np.ndarray = np.linspace(1, 75, 7),
    y_variables: Optional[List[str]] = None,
    x_variables: Optional[List[str]] = None,
    grouping_variables: Optional[List[str]] = None,
    runs: int = 5,
) -> Dict[str, Dict[str, np.ndarray]]:
    """
    Penalized structured Best Linear Unbiased Prediction (psBLUP) method

    Parameters:
    -----------
    Y : pd.DataFrame
        Dependent variables (response matrix)
    X : pd.DataFrame
        Predictor variables (feature matrix)
    proximity_matrix : np.ndarray, optional
        Proximity matrix for structured penalty (default: correlation matrix of X)
    log_transform_x : bool, default False
        Whether to apply log transformation to Y variables
    percent_train_data : float, optional
        Proportion of data to use for training (default: 1.0, no train/test split)
    l2s : np.ndarray
        Array of L2 penalty parameters to test
    y_variables : List[str], optional
        Names of Y variables to analyze (default: all columns in Y)
    x_variables : List[str], optional
        Names of X variables to use (default: all columns in X)
    grouping_variables : List[str], optional
        Variables to remove grouping effects from
    runs : int, default 5
        Number of cross-validation runs

    Returns:
    --------
    Dict[str, Dict[str, np.ndarray]]
        Dictionary with results for each Y variable containing accuracy matrices
    """

    # Arguments validation
    if X is None:
        raise ValueError("No predictor variables have been provided")
    if Y is None:
        raise ValueError("No dependent variables have been provided")

    if proximity_matrix is None:
        proximity_matrix = np.corrcoef(X.select_dtypes(include=[np.number]).T)

    if not isinstance(log_transform_x, bool):
        raise ValueError("Acceptable values for 'log_transform_x': True or False")

    if y_variables is None:
        y_variables = list(Y.select_dtypes(include=[np.number]).columns)

    if x_variables is None:
        x_variables = list(X.select_dtypes(include=[np.number]).columns)

    if percent_train_data is None:
        percent_train_data = 1.0
        runs = 1
        warnings.warn(
            "Since no training data will be used, only 1 run will be conducted."
        )

    objects_to_return = {}

    # Create copies of the dataframes to avoid modifying originals
    X_work = X.copy()
    Y_work = Y.copy()

    # Keep only relevant variables
    if grouping_variables:
        Y_work = Y_work[y_variables + grouping_variables]
    else:
        Y_work = Y_work[y_variables]
    X_work = X_work[x_variables]

    # Rename percentTrainData to proportion
    proportion = (
        [percent_train_data]
        if isinstance(percent_train_data, (int, float))
        else percent_train_data
    )

    # Apply log transformation to the data
    if log_transform_x:
        for var in y_variables:
            Y_work[var] = np.log(Y_work[var])

    # Handle grouping variables
    if grouping_variables:
        Y_work = Y_work[y_variables]

    # Add strength (row-sum) as diagonal (needed for properly defining the Laplacian matrix)
    row_sums = np.sum(proximity_matrix, axis=1)
    np.fill_diagonal(proximity_matrix, row_sums)

    # Scale the X variables
    scaler = StandardScaler()
    X_scaled = pd.DataFrame(
        scaler.fit_transform(X_work), columns=X_work.columns, index=X_work.index
    )

    # Define N as the number of samples
    N = len(X_scaled)

    # Run psBLUP for every metabolite/response variable
    for y_var in tqdm(y_variables, desc="Processing Y variables"):
        print(f"Processing variable: {y_var}")

        # Define response as scaled concentration of the selected variable
        y_scaler = StandardScaler()
        response = y_scaler.fit_transform(Y_work[[y_var]]).flatten()

        # Define predictors as the scaled data
        predictors = X_scaled.copy()

        # Find prediction variables that are negatively correlated to the response
        correlations = []
        for col in x_variables:
            corr, _ = pearsonr(response, predictors[col])
            correlations.append(corr)

        correlations = np.array(correlations)
        negative_correlations = np.where(correlations < 0)[0]
        neg_effects = [x_variables[i] for i in negative_correlations]

        # Flip negatively correlated variables
        for col in neg_effects:
            col_data = predictors[col]
            min_val, max_val = col_data.min(), col_data.max()
            # Map min to max and max to min
            predictors[col] = max_val + min_val - col_data

        # Define the Normalized Laplacian matrix as G
        G = proximity_matrix.copy()
        diag_G = np.diag(G).copy()
        G = -G
        np.fill_diagonal(G, diag_G)

        # Decompose the Normalized Laplacian matrix
        eigenvalues, eigenvectors = linalg.eigh(G)
        eigenvalues[eigenvalues < 0] = 0

        # Define newdata for the augmented data solution
        newdata = (eigenvectors @ np.diag(np.sqrt(eigenvalues))).T

        # Create matrix containing the psBLUP accuracy results
        overall_accuracy = np.zeros((len(proportion), len(l2s)))

        # Create matrix to store accuracy statistics per run
        accuracy_stats = np.full((len(proportion), runs), np.nan)

        # Run accuracy evaluation
        j = 0
        run_progress = tqdm(total=runs, desc=f"Runs for {y_var}", leave=False)
        while j < runs:
            # Create matrix for storing accuracy per sampling scenario
            accuracy_per_scenario = np.full((len(proportion), len(l2s)), np.nan)

            for k in tqdm(
                range(len(proportion)), desc="Train proportions", leave=False
            ):
                # Select training and testing data
                train_indices = np.random.choice(
                    N, size=int(np.round(proportion[k] * N)), replace=False
                )
                test_indices = np.setdiff1d(np.arange(N), train_indices)

                X_train = predictors.iloc[train_indices]
                Y_train = response[train_indices]

                if len(train_indices) == N:
                    X_test = X_train.copy()
                    Y_test = Y_train.copy()
                else:
                    X_test = predictors.iloc[test_indices]
                    Y_test = response[test_indices]

                # Run psBLUP for every l2 value
                for i, l2 in enumerate(tqdm(l2s, desc="L2 values", leave=False)):
                    try:
                        # Calculate factor for augmented data solution
                        factor = (1 + l2) ** (-1 / 2)

                        # Create augmented data for predictors and response
                        X_train_matrix = X_train[x_variables].values
                        augmented_X = factor * np.vstack(
                            [X_train_matrix, np.sqrt(l2) * newdata]
                        )
                        augmented_Y = np.concatenate(
                            [Y_train, np.zeros(newdata.shape[0])]
                        )

                        # Fit psBLUP using linear regression
                        model = LinearRegression(fit_intercept=False)
                        model.fit(augmented_X, augmented_Y)

                        # Obtain psBLUP coefficients
                        coeffs_psblup = model.coef_

                        # Obtain fitted values
                        X_test_matrix = X_test[x_variables].values
                        fitted_psblup = X_test_matrix @ coeffs_psblup

                        # Calculate psBLUP accuracy for specific l2
                        correlation, _ = pearsonr(Y_test, fitted_psblup)
                        accuracy_per_scenario[k, i] = np.round(correlation, 3)

                    except Exception as e:
                        print(f"Error in run {j+1}, proportion {k+1}, l2 {i+1}: {e}")
                        break

                # Store best accuracy for this run
                valid_accuracies = accuracy_per_scenario[
                    k, ~np.isnan(accuracy_per_scenario[k, :])
                ]
                if len(valid_accuracies) > 0:
                    accuracy_stats[k, j] = np.max(valid_accuracies)

            j += 1
            run_progress.update(1)
            overall_accuracy += np.nan_to_num(accuracy_per_scenario, nan=0)

        run_progress.close()
        overall_accuracy = overall_accuracy / j

        objects_to_return[y_var] = {
            "accuracy": overall_accuracy,
            "stats": accuracy_stats,
        }

    return objects_to_return


if __name__ == "__main__":
    pass
